### PR TITLE
Cancel merges on DROP PARTITION

### LINKED ISSUE #17151: TRUNCATE TABLE does not stop already running merges.
And waits for them. That's pointless.

### PR DESCRIPTION
**Pull request by tavplubix**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Cancel already running merges in partition on `DROP PARTITION` and `TRUNCATE` for `ReplicatedMergeTree`. Resolves #17151

### CODE REVIEW

**Review by alesapin — COMMENTED** [review]
Why we cannot cancel them using `merger_mutator.merges_blocker`?

**Comment by tavplubix** [comment]
> Why we cannot cancel them using `merger_mutator.merges_blocker`?

We can, but it will cancel merges in all partitions.

**Comment by alesapin** [comment]
And we don't need the same logic for non-replicated MergeTree because it already stops all merges for each drop partition.

File: src/Storages/StorageReplicatedMergeTree.cpp [hunk 1]
```diff
@@ -7094,6 +7100,8 @@ bool StorageReplicatedMergeTree::dropAllPartsInPartition(
     String log_znode_path = dynamic_cast<const Coordination::CreateResponse &>(*responses.front()).path_created;
     entry.znode_name = log_znode_path.substr(log_znode_path.find_last_of('/') + 1);
 
+    getContext()->getMergeList().cancelInPartition(getStorageID(), partition_id, drop_range_info.max_block);
```
Comment (alesapin) [inline src/Storages/StorageReplicatedMergeTree.cpp:7103]: maybe add some logging here?

**Review by alesapin — APPROVED** [review]

File: src/Storages/StorageReplicatedMergeTree.cpp (hunk 1 above)
Comment (tavplubix) [reply src/Storages/StorageReplicatedMergeTree.cpp:7103]: https://github.com/ClickHouse/ClickHouse/blob/76156af5cc30b8706e2b9527811706cce99a452d/src/Storages/StorageReplicatedMergeTree.cpp#L7062

**Pull request merged**